openlibrary-wikidata-bot/jobs/sync_author_identifiers_from_wikidata.py

```diff
@@ -0,0 +1,210 @@
+logger.addHandler(console_handler)
+# Log INFO+ to the log file
+log_dir = 'logs/jobs/sync_author_wikidata_ids'
+problem_dir = 'exports/jobs/sync_author_wikidata_ids'
+makedirs(log_dir, exist_ok=True)
+makedirs(problem_dir, exist_ok=True)
+log_file_datetime = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+log_file = log_dir + '/sync_author_wikidata_ids_%s.log' % log_file_datetime
+file_handler = logging.FileHandler(log_file)
+file_handler.setLevel(logging.DEBUG)
+file_handler.setFormatter(log_formatter)
+logger.addHandler(file_handler)
+
+# Setup error sheet
+n = datetime.datetime.now().strftime("%Y-%m-%d-%H%M%S")
+csv_file_path = f'{problem_dir}/sync_author_wikidata_ids_merge_problems-{n}.csv'
+with open(csv_file_path, mode="w", newline="", encoding="utf-8") as file:
+    writer = csv.writer(file)
+    writer.writerow(["wdid", "olid", "author_name", "problem", "identifier", "details"])  # Write header
+
+# I'd like it if this could come from identifiers.yml, but that's in a totally different repo!
+# In an ideal world, identifiers.yml stores the WD P### identifier for each remote ID type, and we can loop thru that instead.
+# Wikidata JSONs only use the P### identifier, so that's our only way to target specific remote IDs.
+WD_IDENTIFIERS = {
+    "P214": "viaf",
+    "P2607": "bookbrainz",
+    "P434": "musicbrainz",
+    "P2963": "goodreads",
+    "P213": "isni",
+    "P345": "imdb",
+    "P244": "lc_naf",
+    "P7400": "librarything",
+    "P1899": "librivox",
+    "P1938": "project_gutenberg",
+    "P396": "opac_sbn",
+    "P4862": "amazon",
+    "P12430": "storygraph",
+    "P2397": "youtube"
+}
+
+def validate_wikidata_key(obj: dict[str, any], key=str) -> bool:
+    if not("property" in obj):
+         return False
+    if not ("id" in obj["property"]):
+         return False 
+    if key != obj["property"]["id"]:
+         return False
+    if not ("value" in obj):
+        return False
+    if not ("content" in obj["value"]):
+        return False
+    return True
+
+
+def write_error(wd_id, author_key, author_name, error_type, id, details):
+    logger.error(f'{error_type} for {author_key} ({author_name}), wd {wd_id}: {id}: {details}')
+    with open(csv_file_path, mode="a", newline="", encoding="utf-8") as file:
+        writer = csv.writer(file)
+        writer.writerow([wd_id, author_key, author_name, error_type, id, details])
+
+
+def merge_remote_ids(
+    author, incoming_ids, wd_id
+) -> tuple[dict[str, str], int]:
+    """Returns the author's remote IDs merged with a given remote IDs object, as well as a count for how many IDs had conflicts.
+    If incoming_ids is empty, or if there are more conflicts than matches, no merge will be attempted, and the output will be (author.remote_ids, -1).
+    """
+    if not hasattr(author, "remote_ids"):
+        write_error(wd_id, author.olid, author.name, "openlibrary_author_remote_ids_failed_to_fetch", "", "")
+        return {}, -1
+    output = {**author.remote_ids}
```

Comment:
This should be ok; I'm not sure if the attribute itself will be missing if an author does not have remote_ids. But if that's the case, then that means the author has no remote_ids yet! So we start with an empty dict.

Also in this case it makes no difference since the dictionary is shallow, but to make the code a little more future proof, let's use deepcopy.

```suggestion
    output = deepcopy(getattr(author, 'remote_ids', {}))
```